python/plugins/pydanticai/composio_pydanticai/toolset.py

```diff
@@ -0,0 +1,228 @@
+
+import typing_extensions as te
+from pydantic import ValidationError
+from pydantic_ai.tools import RunContext, Tool
+
+from composio import Action, ActionType, AppType, TagType
+from composio.tools import ComposioToolSet as BaseComposioToolSet
+from composio.tools.toolset import ProcessorsType
+from composio.utils import help_msg
+from composio.utils.shared import get_signature_format_from_schema_params
+
+
+AgentDeps = TypeVar("AgentDeps", bound=Any)
+
+
+@dataclass
+class ToolConfig:
+    """Configuration for a specific tool."""
+
+    max_retries: int = 3
+
+
+class ComposioToolSet(
+    BaseComposioToolSet,
+    t.Generic[AgentDeps],
+    runtime="pydantic_ai",
+    description_char_limit=1024,
+    action_name_char_limit=64,
+):
+    """
+    Composio toolset for Pydantic-AI framework.
+    Example:
+    ```python
+    from dotenv import load_dotenv  # type: ignore
+    import os
+
+    from composio import Action
+    from composio_pydanticai import ComposioToolSet
+    from pydantic_ai import Agent
+
+
+    # Load environment variables from .env
+    load_dotenv(".env")
+    # Initialize tools with custom max retries
+    composio_toolset = ComposioToolSet(max_retries=5)  # Configure max retries for all tools
+
+    # Get GitHub tools that are pre-configured
+    tools = composio_toolset.get_tools(
+        actions=[Action.GITHUB_STAR_A_REPOSITORY_FOR_THE_AUTHENTICATED_USER]
+    )
+
+    # Create an agent with the tools
+    agent = Agent(
+        model="openai:gpt-4o",  # Using a known model name
+        tools=tools,
+        system_prompt="You are an AI agent that helps users interact with GitHub.
+        You can perform various GitHub operations using the available tools.
+        When given a task, analyze it and use the appropriate tool to complete it.",
+    )
+
+    # Define task
+    task = "Star a repo wjayesh/mahilo on GitHub"
+    # Run the agent
+    result = agent.run_sync(task)
+    ```
+    """
+
+    def __init__(
+        self,
+        max_retries: int = 3,
+        tool_configs: t.Optional[Dict[str, ToolConfig]] = None,
+        *args,
+        **kwargs,
+    ):
+        """Initialize the toolset with configurable max_retries.
+
+        Args:
+            max_retries (int, optional): Default maximum number of retries for tool execution. Defaults to 3.
+            tool_configs (Dict[str, ToolConfig], optional): Specific configurations for individual tools.
+                Keys are tool names, values are ToolConfig objects.
+        """
+        kwargs["max_retries"] = max_retries
+        super().__init__(*args, **kwargs)
+        self.tool_configs = tool_configs or {}
+
+    def _wrap_action(
+        self,
+        action: str,
+        description: str,
+        schema_params: t.Dict[str, t.Any],
+        entity_id: t.Optional[str] = None,
+    ):
+        """Create a wrapper function for the Composio action."""
+
+        async def function(
+            ctx: "RunContext[AgentDeps]", **kwargs: t.Any
```

Comment:
This does not mean anything unless the type annotations are delegated outside of the wrapper function